### ISSUE TITLE
Keyboard behaviour of custom select options

### ISSUE THREAD

**Issue by o-t-w** [openui/open-ui#1465]
In Safari all the arrow keys work to move focus. In Chrome only the up and down buttons work. 

The layout of the options of a custom select may be horizontal, vertical, or in a grid. When the layout is not vertical, the behaviour in Chrome is problematic because the up/down keys have to be used to move left and right. When the options have a grid layout, neither the Chrome nor Safari behaviour is ideal.

**Comment by josepharhar** [comment]
Thanks for the feedback! We could make right arrow and down arrow always go forward, and make left and up always go backward. That is what safari is doing, right?

As for grid, there was a related issue here: https://github.com/openui/open-ui/issues/1229

It kind of sounds like spatial navigation is what people want for that? I wonder if spatial navigation with arrow keys has any other use cases besides select or had any other proposals to actually be web exposed

**Comment by scottaohara** [comment]
2-dimensional/spacial navigation would definitely need some sort of a11y api mapping so that screen readers could communicate the fact the listbox no longer behaves like other listboxes.   Or, maybe it could somehow be re-mapped as a grid.... that would at least communicate the 2d layout./ keyboarding expectations,..  though imo it'd be better to just make that (grid) a type of supported popup, rather than have the a11y mappings change based on styling choices

**Comment by lukewarlow** [comment]
The other proposals would essentially be 3d focus groups which is definitely something we want but is also something that needs more thoughts.

**Comment by gregwhitworth** [comment]
Looping in the ARIA group

**Comment by josepharhar** [comment]
> Or, maybe it could somehow be re-mapped as a grid.... that would at least communicate the 2d layout./ keyboarding expectations,.. though imo it'd be better to just make that (grid) a type of supported popup, rather than have the a11y mappings change based on styling choices

Grid could be a whole separate primtive that we make a new html element for, right? I'm not super familiar with grids, but maybe that would be a good path to go down to support this somehow?

I think that for now we should decide on whether the left/right arrows should go back/forward or not in horizontal-tb, and whether up/down should go back/forward in vertical writing modes, which is what safari is doing based on my reading of the OP.

**Comment by brechtDR** [comment]
So... there are a few things to consider here. When (certain) people go wild with styling, there are a few things that can happen. I'd like to add a bit more info.

Here I have a multi-select, a grid would make sense here in some level:
https://codepen.io/utilitybend/pen/xbOOayg

However, since we are in a select, I don't find this particularity needed. I'll leave this open to the group, but wanted to highlight an example.

But the left and right buttons can have some downsides:
https://codepen.io/utilitybend/pen/poMpNbg

In this demo, I added a horizontal scroller with scroll snapping. I noticed this demo did not work on Safari, which is a shame. I know this is already highly customized and a gamified experience (which you shouldn't use on a regular website, but maybe for a browser game?)

Still, it feels like it's fighting with potential examples, and it should be specified whether these arrows do something, or not, and preferably be the same across each browser.

**Comment by josepharhar** [comment]
Here's a demo which shows the different keyboard behavior we implemented:
```html
<style>
select, ::picker(select) {
  appearance: base-select;
}
</style>

<select>
  <option>one</option>
  <option>two</option>
  <option>three</option>
</select>

<select style="writing-mode:vertical-lr">
  <option>one</option>
  <option>two</option>
  <option>three</option>
</select>

<select style="writing-mode:vertical-rl">
  <option>one</option>
  <option>two</option>
  <option>three</option>
</select>
```

Putting the behaviors in a table feels hard so I'll just say it:
- For vertical-lr, chrome and safari do the same thing: left and up arrow keys go backwards, right and down arrow keys go forwards.
- For vertical-rl, chrome and safari do the same thing: right and up arrow keys go backwards, left and down arrow keys go forwards.
- For horizontal-tb, chrome and safari are different:
  - chrome: up arrow goes backwards, down arrow goes forwards. left and right keys do nothing.
  - safari: left and up arrows go backwards, right and down arrows go forwards. same as vertical-lr.

I implemented the chrome behavior to simply match the safari behavior which they put in a wpt here, which must have not tested right/left in horizontal-tb: https://github.com/web-platform-tests/wpt/commit/d2ad58e1ff0ac60de95ebeb937bb9a70b4e08852

**Comment by LeaVerou** [comment]
> I wonder if spatial navigation with arrow keys has any other use cases besides select or had any other proposals to actually be web exposed

Any focusable control with interactive grid children. E.g. a data grid also needs it.

And it's impossibly hard to implement well in the general case (where you're unaware of the layout characteristics besides what gCS can provide).
